File: binarytree/144_preorder_traversal.cc
#include <stack>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> ret;
        if (root == nullptr)
        {
            return ret;
        }

        stack<TreeNode*> _stack {{root}};

        while (!_stack.empty())
        {
            TreeNode* node = _stack.top();
            ret.push_back(node->val);
            _stack.pop();

            if (node->right != nullptr)
            {
                _stack.push(node->right);
            }
            if (node->left != nullptr)
            {
                _stack.push(node->left);
            }
        }
        return ret;
    }
};

int main() {
    TreeNode* root = new TreeNode(1);
    root->right = new TreeNode(2);
    root->right->left = new TreeNode(3);
    vector<int> ret = Solution().preorderTraversal(root);
    for_each(ret.begin(), ret.end(), [](int &i){ printf("%d\n", i); });
    return 0;
}

File: binarytree/144_preorder_traversal.py
from typing import List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# class Solution:
#     def dfs(self, root: TreeNode, nums: List[int]):
#         if not root:
#             return
#         nums.append(root.val)
#         if root.left:
#             self.dfs(root.left, nums)
#         if root.right:
#             self.dfs(root.right, nums)

#     def preorderTraversal(self, root: TreeNode) -> List[int]:
#         nums: List[int] = []
#         self.dfs(root, nums)
#         return nums

# class Solution:

#     def preorderTraversal(self, root: TreeNode) -> List[int]:
#         if not root:
#             return []
#         return [
#             root.val,
#             *self.preorderTraversal(root.left),
#             *self.preorderTraversal(root.right)
#         ]

# class Solution:

#     def preorderTraversal(self, root: TreeNode) -> List[int]:
#         if not root:
#             return []
#         nums: List[int] = []
#         stack: List[TreeNode] = [root]
#         while stack:
#             node = stack.pop()
#             nums.append(node.val)
#             if node.right:
#                 stack.append(node.right)
#             if node.left:
#                 stack.append(node.left)
#         return nums

class Solution:

    def preorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []
        nums: List[int] = []
        stack: List[TreeNode] = []
        curr: TreeNode = root
        while curr or stack:
            if curr:
                stack.append(curr)
                nums.append(curr.val)
                curr = curr.left
            else:
                node: TreeNode = stack.pop()
                curr = node.right
        return nums




if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    print(Solution().preorderTraversal(root))